### PR TITLE
PERF-2689 New benchmarks for large datasets with indexes

### PR DESCRIPTION
**Pull request by IrinaYatsenko**
Evergreen runs:
https://spruce.mongodb.com/version/618d50b45623435d06bc1b3e/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC
https://spruce.mongodb.com/version/618d50172fbabe51bb1beeb1/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

### CODE REVIEW

File: testcases/simple_query.js [hunk 1]
```diff
@@ -1036,4 +1036,48 @@ if (typeof(tests) !== "object") {
...
+        indexes: [{"a":1}, {"b":1}, {"a":1, "b":1}],
+        query: {"a": 7, "b":742}
+    });
+    addTestCaseWithLargeDatasetAndIndexes({
+        name: "PointQuerySubField_SingleIndex_LL", 
+        docGenerator: largeDoc,
+        indexes: [{"e.a":1}],
+        query: {"e.a": 7}
+    });
+    addTestCaseWithLargeDatasetAndIndexes({
```
Comment (jimoleary) [inline testcases/simple_query.js:1077]: Very much an edge case but do we also want an embedded document index test?

```
addTestCaseWithLargeDatasetAndIndexes({
        name: "PointQuerySubFields_EmbeddedMultipleIndexes_LL", 
        docGenerator: largeDoc,
        indexes: [{"e.a":1}, {"e.b":1}, {"e":1}],
        query: {e:{a: 7, b:742}}
    });
```

File: testcases/pipelines.js [hunk 2]
```diff
@@ -2343,5 +2343,77 @@ generateTestCaseWithLargeDataset({
...
+    options.pre = function(collection) {
+        dropIndexesAndCaches(collection);
+        createIndexes(collection, options.indexes);
+    };
+    options.post = dropIndexesAndCaches;
+    generateTestCaseWithLargeDataset(options);
+}
+
+// Grouping on the indexed field with no accumulators creates DISTINCT_SCAN plan with a
+// $groupByDistinctScan stage, that currently isn't being lowered into SBE.
```
Comment (jimoleary) [inline testcases/pipelines.js:2367]: Should ```lowered``` be ```pushed down```? The scope, design and SERVER-58429 use the term pushdown.

**Review by jimoleary — COMMENTED** [review]
I have some minor questions but the works looks pretty good.

File: testcases/simple_query.js (hunk 1 above)
Comment (IrinaYatsenko) [reply testcases/simple_query.js:1077]: With these benchmarks I'm trying to pinpoint specific components of the execution engine performance, especially in the context of comparing the old engine to the new one. The "PointQuery_SingleIndex_LL" case covers having IXSCAN & FETCH stages, "PointQuery_MultipleIndexes_LL" triggers multi-planning with the associated overhead and we are curious to know how big that overhead is. And the "SubFields" cases are to provide comparison with the corresponding non-indexed ones. I'm actually considering removing "PointQuerySubFields_MultipleIndexes_LL" because it doesn't provide any additional signal. I believe, the test case with embedded document index falls into the same bucket.

File: testcases/pipelines.js [hunk 3]
```diff
@@ -2343,5 +2343,77 @@ generateTestCaseWithLargeDataset({
...
+generateTestCaseWithLargeDatasetAndIndexes({
```
[the remark this answers is not in the document]
Comment (IrinaYatsenko) [reply testcases/pipelines.js:2368]: Hm. Even when there is an index that covers both group_by field and the accumulator field, we generate a COLLSCAN (no indexes are considered). While this is a valid scenario and potentially something to optimize, currently this case would be the same as Group.SumAccTopField_LL10. I'm worried to grow this test suite too much so I'll leave this case out until we specifically target it (the benchmarks being added here are to support the work in SERVER-58429).

**Review by IrinaYatsenko — COMMENTED** [review]
Replied to @jimoleary comments. Jim, please let me know if you feel strongly about any of them.

File: testcases/pipelines.js (hunk 2 above)
Comment (jimoleary) [reply testcases/pipelines.js:2367]: kk

File: testcases/pipelines.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jimoleary) [reply testcases/pipelines.js:2368]: makes sesnse, I was just asking for the sake of completeness.

File: testcases/simple_query.js (hunk 1 above)
Comment (jimoleary) [reply testcases/simple_query.js:1077]: understood, it's very much an edge case anyway.

**Review by jimoleary — APPROVED** [review]

**Comment by IrinaYatsenko** [comment]
@jimoleary Jim, could you please take another look? I've ran the benchmarks in Evergreen and realized that we are missing signal from low selectivity queries, so I've added them (and will kick off another set of Evergreen runs). Thanks!

File: testcases/simple_query.js [hunk 4]
```diff
@@ -1036,4 +1036,60 @@ if (typeof(tests) !== "object") {
...
+    }
+
+    addTestCaseWithLargeDatasetAndIndexes({
+        name: "PointQuery_SingleIndex_LL", 
+        docGenerator: largeDoc,
+        indexes: [{"a":1}],
+        query: {"a": 7}
+    });
+    addTestCaseWithLargeDatasetAndIndexes({
+        name: "PointQuery_SingleIndex_LowSelectivityMatch_LL",
```
Comment (jimoleary) [inline testcases/simple_query.js:1066]: nit: shouldn't the name be ```RangeQuery_SingleIndex_LowSelectivityMatch_LL``` (same for all the $gt queries).

**Review by jimoleary — APPROVED** [review]
lgtm but left a question about test names.

File: testcases/pipelines.js [hunk 5]
```diff
@@ -2343,5 +2343,73 @@ generateTestCaseWithLargeDataset({
...
+    name: "Group.NoAccTopField_DistinctScan_LL10", 
+    docGenerator: largeDoc,
+    indexes: [{"a":1}],
+    pipeline: [{$group: {_id: "$a"}}]
+});
+
+// The $match stage triggers IXSCAN plans, without it, the only considered plan is COLLSCAN.
+// The filter might considerably reduce the size of the data the aggregation runs on, so we try with
+// filters that have artifitially low selectivity as well.
+generateTestCaseWithLargeDatasetAndIndexes({
```
Comment (ericox) [inline testcases/pipelines.js:2378]: These look good to me.

**Review by ericox — APPROVED** [review]
LGTM from the query side.
